### PR TITLE
style(*): update code style to org guidelines

### LINKED ISSUE #164: Update code-style
Get if from https://github.com/posthtml/posthtml-plugin-boilerplate

### PR DESCRIPTION
**Pull request by michael-ciniawsky**
Fixes #164 :)
Close #166

### CODE REVIEW

**Comment by voischev** [comment]
Coverage decreased -2.0% why?

**Comment by michael-ciniawsky** [comment]
¯\_(ツ)_/¯ i was wondering myself, no logic changes from my side, it' those two lines (108, 109)

```js
    if (err) {
      cb(err) // 108
      return  // 109
    }
```
 If someone can guide me where in `test/` this should by tested i try to fix it, albeit i started to migrate them to AVA anyways :), but that's a separate PR

File: lib/index.js [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+ */
+function tryCatch (tryFn, catchFn) {
+  try {
+    return tryFn()
+  } catch (err) {
+    catchFn(err)
+  }
+}
+
+function extend (tree) {
```
Comment (voischev) [inline lib/index.js]: return apiExtend name
& maybe make jsdoc?

File: lib/index.js [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+}
+
+/**
+ * Wraps tree in object with getter that renders it
+ * as HTML on demand.
+ *
+ * @param {PostHTMLTree} tree - json tree to wrap
+ * @returns {html: String, tree: PostHTMLTree} - result
+ */
+function lazy (render, tree) {
```
Comment (voischev) [inline lib/index.js]: lazyResult

File: lib/index.js (hunk 1 above)
Comment (michael-ciniawsky) [inline lib/index.js]: kk , ```lazyRender```  was also renamed to ```lazy``` ;)

**Comment by voischev** [comment]
upd files
- .npmignore (close https://github.com/posthtml/posthtml/pull/166)
- .editorconfig
- .gitignore?

File: lib/index.js (hunk 2 above)
Comment (michael-ciniawsky) [inline lib/index.js]: kk

**Comment by michael-ciniawsky** [comment]
.npmignore -(.jscsrc, .jshintrc, .jshintignore), +(docs, MAINTAINERS, .editorconfig) ?
.editorconfig indent_size = 2 right?
.gitignore What specific changes there?

File: lib/index.js [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+ * @param {?Function} options.parser - alternate html parser to replace default
+ * @param {?Boolean} options.sync - to run plugins synchronously, will throw
+ *                                  if there are async plugins
+ * @returns {Promise<{html: String, tree: PostHTMLTree}>|
+ *          {html: String, tree: PostHTMLTree}} - result
+ */
+PostHTML.prototype.process = function (tree, options) {
+  options = options || {}
+
+  if (options.parser) parser = options.parser
```
Comment (voischev) [inline lib/index.js]: make doc https://github.com/posthtml/posthtml/blob/master/docs/api.md#options

File: lib/index.js [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+  if (options.render) render = options.render
```
Comment (voischev) [inline lib/index.js]: make doc https://github.com/posthtml/posthtml/blob/master/docs/api.md#options

Comment (michael-ciniawsky) [inline lib/index.js]: 👍

File: lib/index.js (hunk 3 above)
Comment (michael-ciniawsky) [inline lib/index.js]: 👍

**Comment by awinogradov** [comment]
LGTM

**Comment by awinogradov** [comment]
ping

**Comment by michael-ciniawsky** [comment]
[Core Docs]https://github.com/michael-ciniawsky/posthtml/blob/code-style/docs/core.md
[API Docs]https://github.com/michael-ciniawsky/posthtml/blob/code-style/docs/api.md

```bash
npm run docs:core
npm run docs:api
```

**Comment by awinogradov** [comment]
@michael-ciniawsky 👍 

But! Can you return 100%?))

**Comment by michael-ciniawsky** [comment]
yep i try, but it's weird that it decreased...

File: .npmignore [hunk 5]
```diff
@@ -1,8 +1,18 @@
...
+# Files
+.editorconfig
 .gitignore
+.mversionrc
 .travis.yml
+
 npm-debug.log
-logo.svg
 
+README.md
```
Comment (voischev) [inline .npmignore]: should stay

File: README.md [hunk 6]
```diff
@@ -1,89 +1,98 @@
...
-[![Chat][chat]][chat-badge]
+[![NPM][npm]][npm-url]
+[![Deps][deps]][deps-url]
+[![Tests][build]][build-url]
+[![Coverage][cover]][cover-url]
+[![Standard Code Style][style]][style-url]
+[![Chat][chat]][chat-url]
 
-# PostHTML <img align="right" width="220" height="200" title="PostHTML logo" src="http://posthtml.github.io/posthtml/logo.svg">
+# PostHTML <img align="right" width="200" height="220" title="PostHTML" src="http://posthtml.github.io/posthtml/logo.svg">
```
Comment (voischev) [inline README.md]: ?))

File: lib/index.js [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+
+var parser = require('posthtml-parser')
+var render = require('posthtml-render')
+
+/**
+ * @author Ivan Voischev (@voischev),
+ *         Anton Winogradov (@awinogradov),
+ *         Alexej Yaroshevich (@zxqfox),
+ *         Vasiliy (@Yeti-or)
+ * @version 0.9.0
```
Comment (voischev) [inline lib/index.js]: oh. automize update this? or delete line

File: lib/index.js [hunk 8]
```diff
@@ -0,0 +1,247 @@
+var api = require('./api.js')
+
+var parser = require('posthtml-parser')
+var render = require('posthtml-render')
+
+/**
+ * @author Ivan Voischev (@voischev),
```
Comment (voischev) [inline lib/index.js]: not necessarily these lines :)

File: README.md [hunk 9]
```diff
@@ -180,103 +192,67 @@ module.exports = {
...
 
 ## Middleware
 
-<img align="right" height="56" title="KoaJS" src="http://t2.gstatic.com/images?q=tbn:ANd9GcRfnGHcTYGyMNcicU4N3nzV-5Rta9s_e5LzSI2HBjKMsLHundtmqAlQ" />
-
-### [Koa](http://koajs.com/)
+### [Koa](https://github.com/posthtml/koa-posthtml) <img align="right" height="56" title="Koa" src="http://t2.gstatic.com/images?q=tbn:ANd9GcRfnGHcTYGyMNcicU4N3nzV-5Rta9s_e5LzSI2HBjKMsLHundtmqAlQ" />
 
-##### [koa-posthtml](https://github.com/michael-ciniawsky/koa-posthtml)
+### [Hapi](https://github.com/posthtml/hapi-posthtml) <img align="right" height="56" title="Hapi" src="https://worldvectorlogo.com/logos/hapi.svg" />
```
Comment (voischev) [inline README.md]: all images looks not good :(

Comment (voischev) [inline README.md]: maybe delete logos?

**Comment by voischev** [comment]
Awesome!

File: README.md (hunk 9 above)
Comment (awinogradov) [inline README.md]: +1

Comment (voischev) [inline README.md]: in NPM markdown

and 



looks not good too

File: .npmignore (hunk 5 above)
Comment (michael-ciniawsky) [inline .npmignore]: 👍

File: README.md (hunk 9 above)
Comment (michael-ciniawsky) [inline README.md]: yep either delete , or trying if it works with ```align=rigth``` removed. Is there a way/service/site to test npm readme preview? :)

File: README.md (hunk 6 above)
Comment (michael-ciniawsky) [inline README.md]: aligned to logo dimensions as in plugin boilerplate, but could be the other way around  aswell :)

**Comment by michael-ciniawsky** [comment]
crop the plugins table to plugins maintained in the org / or rm it completely and make the link to the catalog a more prominent ? I talk with maltsev to iterate the plugins catalog  a little when i find the time :)

File: README.md (hunk 9 above)
Comment (michael-ciniawsky) [inline README.md]: text

```hspace="50%"``` should move to the left like in the screenshot above, for github preview only 

text_hspace

logo

```hspace="50%"``` should move to the left like in the screenshot above, for github preview only 

logo_hspace

Choose what you like, or if disliked in general we use just the headings :)

Comment (voischev) [inline README.md]: NPM clips your styles in markdown ;)

**Comment by voischev** [comment]
ok

File: README.md (hunk 9 above)
Comment (voischev) [inline README.md]: for NPM we can crop images in README automatically... but it seems unnecessary

Comment (michael-ciniawsky) [inline README.md]: so which one of the above do you prefer? :)

File: README.md (hunk 6 above)
Comment (voischev) [inline README.md]: check plz

File: README.md (hunk 9 above)
Comment (voischev) [inline README.md]: delete logos :(

File: README.md (hunk 6 above)
Comment (michael-ciniawsky) [inline README.md]: ? :) plugin boilerplate has ```{width: 200 height: 220}```, i changed it solely for that reason, but other way round i meant we could use ```{width: 220, height: 200)``` in posthtml-plugin-boilerplate if it should stay as s here.

File: README.md (hunk 9 above)
Comment (michael-ciniawsky) [inline README.md]: ok, why not logos only instead of text headings like in 3.)? should not cause any trouble on npm

File: README.md (hunk 6 above)
Comment (voischev) [inline README.md]: i'm check it
return `width: 220, height: 200` and fix in posthtml-plugin-boilerplate plz

)

File: README.md (hunk 9 above)
Comment (voischev) [inline README.md]: in 3 is not a markdown doc style for tech project %)

Comment (michael-ciniawsky) [inline README.md]: kk

**Comment by michael-ciniawsky** [comment]
plugins list rm or crop to org plugins? :)

**Comment by michael-ciniawsky** [comment]
add custom parser && render example or later ?

File: README.md (hunk 6 above)
Comment (michael-ciniawsky) [inline README.md]: yep :), my bad here 👍

**Comment by voischev** [comment]
> plugins list rm or crop to org plugins? :)

stay now

> add custom parser && render example or later ?
now

**Comment by michael-ciniawsky** [comment]
> stay now

Keep the hole plugin list as is, a little more concrete here please :)?

> now

kk. How should custom parser/render modules be named/prefixed by convention?

```
posthtml-parser-[name]
posthtml-render-[name]
posthtml-* (no distinction to plugins)
```

**Comment by voischev** [comment]
@michael-ciniawsky 
stay plugin list as is

> posthtml-* (no distinction to plugins)

like

**Comment by michael-ciniawsky** [comment]
kk, i'm finishing it tomorrow

**Comment by voischev** [comment]
@michael-ciniawsky 👍 is good!

**Comment by voischev** [comment]
@michael-ciniawsky ping

**Comment by michael-ciniawsky** [comment]
@voischev sry incoming in the next few hours :), currently on my way finishing posthtml-hbs(parser), posthtml-pug(parser), posthtml-js(render), poshtml-jsx(render) they at the respective  and render section in the README, the timing was a little unfortunate tbh

**Comment by voischev** [comment]
kk

**Comment by voischev** [comment]
mb remove approvals/lgtm ?

**Comment by michael-ciniawsky** [comment]
sry for the delay 😛  :shipit:

**Comment by michael-ciniawsky** [comment]
I'm on my way to Rome this weekend, if minor takes place, please 

```js
PostHTML.prototype.process = function (tree, options) {
   ...
   render = options.render || render
   ...
}
```

**Comment by voischev** [comment]
>  if minor takes place, please
i don't understand

**Comment by michael-ciniawsky** [comment]
:) when you bump version (patch 0.9.1), because of fixing docs on npm etc., maybe minor release with custom renderer option like i posted above for js jsx export/support? 😛 I only need to finish testing for these, but i'm back to normal working days now and finish it soon.

```html
<div class=${style.app}>${app.content}</div>
```

```js

const options = {
  name: 'Test',
  props: ['one', 'two'],
  type: 'es2015',
  export: true
}

posthtml()
  .process(html, { render: require('posthtml-jsx')(options) }) // Custom Render (JSX Stringifier)
  .then((result) => console.log(result.html))
```

```js 
import {Component} from 'react'

class Test extends Component({
   render (one, two) {
     return (
       <div className={style.app}>{app.content}</div>
    )
  }
})

export default Test
```

like htmltojsx from React community but with more options (types: ['stateless', 'es5', 'es2015', 'jsx' /* Pure JSX for Vue.js and friends, no react */])

**Comment by voischev** [comment]
waiting you. ;) and release v0.10.0. OK?

**Comment by michael-ciniawsky** [comment]
@voischev yep, if README is basically ok for now patch(0.9.1) and when custom render(s) are ready discuss if next minor(0.10.0) ships with this option (options.render), and maybe options.file[name]  ``` => tree.file = { path: dirname, name: 'index.html' }``` (for example) aswell, as the support channel shows, things like includes/modules etc. are always kind of troublesome to setup for beginners, with option.file it would be possible to access the ```file.path``` within gulp-posthtml, posthtml-loader & friends propagate it to posthtml and plugin options related to file paths can be static.

**current**
```js
import tap from 'gulp-tap'
import posthtml from 'gulp-posthtml'
import { task, src, dest } from 'gulp'

task('html', () => {
  let path
  const plugins = [ require('posthtml-include')(root: `${path}`) ]
  const options = {}

  src('src/**/*.html')
    .pipe(tap((file) => path = file.path))
    .pipe(posthtml(plugins, options))
    .pipe(dest('build/'))
})
```

**with options.file**
```js
import { task, src, dest } from 'gulp'
import posthtml from 'gulp-posthtml'

task('html', () => {
  const plugins = [ require('posthtml-include')({ root: 'components' }) ] // relative path
  const options = {}

  src('src/**/*.html')
    .pipe(posthtml(plugins, options)) // access file.path in gulp-posthtml
    .pipe(dest('build/'))
})
```

**Comment by michael-ciniawsky** [comment]
anyways if README approved

:shipit: 

Ship It!

😛

**Comment by voischev** [comment]
ok. merge

**Comment by michael-ciniawsky** [comment]
👍

**Pull request merged**